techmatters/hrm
# feat: Client Profiles permissions

## Description
This PR adds support for profiles and profile sections to our permissions framework:
- `profile`
  - Supports the following actions:
    - `viewProfile`: grants permission to load a single profile details.
    - `flagProfile`: grants permission to associate a flag to a profile.
    - `unflagProfile`: grants permission to disassociate a flag to a profile.
  - The valid predicates on profiles actions are:
    - `TimeBasedCondition` (`createdHoursAgo`, `createdDaysAgo`).
    - `UserBasedCondition` (`isSupervisor`, `everyone`).

- `profileSection` 
  - Supports the following actions:
    - `createProfileSection`: grants permission to create a new profile section.
    - `viewProfileSection`: grants permission to view a single profile section.
    - `editProfileSection`: grants permission to edit a single profile section.
  - The valid predicates on profile sections actions are:
    - `TimeBasedCondition` (`createdHoursAgo`, `createdDaysAgo`).
    - `UserBasedCondition` (`isSupervisor`, `everyone`).
    - `ProfileSectionSpecificCondition`: for now the only valid predicate here is `sectionType`. Similarly to the time based ones, this predicate expects a parameter, the section type, to then evaluate against an actual section. That means that the predicate will be represented in an object format, like `{ "sectionType": "some-section" }` - this would mean "if the section type is `"some-action"`. Example:
      - `[["isSupervisor"], [{ "sectionType": "summary" }]]`, which is the same as `[["isSupervisor"], ["everyone", { "sectionType": "summary" }]]` means that the action can be performed only if the user is a supervisor or if the section is of type `summary`.


### Checklist
- [x] [Corresponding issue has been opened](https://tech-matters.atlassian.net/browse/CHI-2369)

### Verification steps
- Change `hrm-domain/hrm-core/permission-rules/dev.json` profile permissions like
  ``` 
  "viewProfile": [["everyone"]],
  "flagProfile": [["everyone"]],
  "unflagProfile": [["everyone"]],
  "createProfileSection": [["everyone"]],
  "viewProfileSection": [["everyone"]],
  "editProfileSection": [["everyone"]],
  ```
- Build (`npm run build`) and run local server (`npm run start:light`)
- Run Flex locally and point to the local server (`npm run dev:local` from Flex plugin).
- Confirm that you are able to interact with all the parts related to profiles as usual:
  -  Open profile details.
  - Associate and disassociate profile flags.
  - Create, view and edit any profile section.
- Change `hrm-domain/hrm-core/permission-rules/dev.json` profile permissions like
  ``` 
  "viewProfile": [ ],
  "flagProfile": [ ],
  "unflagProfile": [ ],
  "createProfileSection": [ ],
  "viewProfileSection": [ ],
  "editProfileSection": [ ],
  ```
- Repeat the above steps and confirm you can't perform any of the above actions.

NOTES: 
- You will be able to "see profile details" because there is already partially loaded information in the state - retrieved from the "list profiles" action. We should handle the "rejection" state in the UI, since the backend service is already sending the corresponding error code.
- Since the "search case/contact by profile id" endpoints are implemented using the same underlying code as the "search endpoints", the permissions should work in the same manner.
